lib/evaluators/AsyncEvaluator.ts

```diff
@@ -1,23 +1,27 @@
 import * as RDF from 'rdf-js';
-import { Algebra as Alg } from 'sparqlalgebrajs';
+import {Algebra as Alg} from 'sparqlalgebrajs';
 
 import * as E from '../expressions/Expressions';
 
-import { transformAlgebra } from '../Transformation';
-import { Bindings, ExpressionEvaluator } from '../Types';
+import {transformAlgebra} from '../Transformation';
+import {Bindings, ExpressionEvaluator} from '../Types';
 
-import { AsyncRecursiveEvaluator } from './RecursiveExpressionEvaluator';
+import {AsyncRecursiveEvaluator} from './RecursiveExpressionEvaluator';
 
 type Expression = E.Expression;
 type Term = E.TermExpression;
 
+export type AsyncExtensionFunction = (args: RDF.Term[]) => Promise<RDF.Term>;
+export type AsyncExtensionFunctionCreator = (functionNamedNode: RDF.NamedNode) => AsyncExtensionFunction | null;
```

Comment:
Just wondering, would the following be (too) complicated to support?
```suggestion
export type AsyncExtensionFunctionCreator = (functionNamedNode: RDF.NamedNode) => Promise<AsyncExtensionFunction | null>;
```

Reply:
Also, can we use `undefined` everywhere instead of `null`, as explained in the Comunica PR?

Reply:
I don't think it would be, not sure. I'm not sure why you would want this. Anyway I see to ways in handeling this.
1. We make the constructor async and everything else stays the same.
2. In transform named we call the creator function. This gives us a promise which we await in the `evalAsyncExtension` function. Problem here is that when the promise revolves to be null/ undefined, we would need to throw the `UnknownNamedOperator` error or something, this behavior is completely different from what we have now. Right now we expect that we know whether we can execute an expression on the creation of the evaluator. This brings us to the biggest problem. Libraries like comunica use the creator as a test to check if a query is executable. This assertion would not be valid anymore.

Reply:
Ok, let's leave it for future work then.

Reply:
looking at the first suggestion I made. It's impossible to make a constructor async. We would need something like `async readyAsyncEvaluator()` to handle this than. What are your thoughts on this? When we do this we should not forget to edit the README.

Reply:
Let's just leave it out for now then, It's not that important to have for now :-)

Reply:
I just thought of another solution. We could make a hybrid of 1 and 2. We can do everything 2 would do and just provide something like `async readyEvaluation` which awaits the promise (NOTE that calling this function would be optional). Comunica might than be able to call this function in it's test. It shouldn't be to hard to implement this.

Reply:
Seems too error-prone, would have to be something mandatory.
But I wouldn't worry about it too much now, we can revisit it via a separate PR if needed.

Reply:
Alright! I'll resolve the conversation than.